cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -293,6 +303,90 @@ Status DiscoverFilesFromDir(const std::shared_ptr<fs::LocalFileSystem>& local_fs
+                          FromProto(agg_func, is_hash, ext_set, conversion_options));
+    ExtensionIdRegistry::SubstraitAggregateToArrow converter;
+    if (aggregate_call.id().uri.empty() || aggregate_call.id().uri[0] == '/') {
+      ARROW_ASSIGN_OR_RAISE(converter,
+                            ext_set.registry()->GetSubstraitAggregateToArrowFallback(
+                                aggregate_call.id().name));
+    } else {
+      ARROW_ASSIGN_OR_RAISE(converter, ext_set.registry()->GetSubstraitAggregateToArrow(
+                                           aggregate_call.id()));
+    }
+    ARROW_ASSIGN_OR_RAISE(compute::Aggregate arrow_agg, converter(aggregate_call));
+
+    // find aggregate field ids from schema
+    const auto& target = arrow_agg.target;
+    size_t measure_id = agg_src_fieldsets.size();
+    agg_src_fieldsets.push_back({});
+    for (const auto& field_ref : target) {
+      ARROW_ASSIGN_OR_RAISE(auto match, field_ref.FindOne(*input_schema));
+      agg_src_fieldsets[measure_id].push_back(match[0]);
+    }
+
+    aggregates.push_back(std::move(arrow_agg));
+    return Status::OK();
+  } else {
+    return Status::Invalid("substrait::AggregateFunction not provided");
+  }
+}
+
+ARROW_ENGINE_EXPORT Result<DeclarationInfo> MakeAggregateDeclaration(
+    std::optional<substrait::RelCommon> agg_common_opt, compute::Declaration input_decl,
+    std::shared_ptr<Schema> input_schema, const int measure_size,
+    std::vector<compute::Aggregate> aggregates,
+    std::vector<std::vector<int>> agg_src_fieldsets, std::vector<FieldRef> keys,
+    std::vector<int> key_field_ids, std::vector<FieldRef> segment_keys,
+    std::vector<int> segment_key_field_ids, const ExtensionSet& ext_set,
+    const ConversionOptions& conversion_options) {
+  FieldVector output_fields;
+  output_fields.reserve(key_field_ids.size() + segment_key_field_ids.size() +
+                        measure_size);
+  // extract aggregate fields to output schema
+  for (const auto& agg_src_fieldset : agg_src_fieldsets) {
+    for (int field : agg_src_fieldset) {
+      output_fields.emplace_back(input_schema->field(field));
+    }
+  }
+  // extract key fields to output schema
+  for (int key_field_id : key_field_ids) {
+    output_fields.emplace_back(input_schema->field(key_field_id));
+  }
+  // extract segment key fields to output schema
+  for (int segment_key_field_id : segment_key_field_ids) {
+    output_fields.emplace_back(input_schema->field(segment_key_field_id));
+  }
+
+  std::shared_ptr<Schema> aggregate_schema = schema(std::move(output_fields));
+
+  DeclarationInfo aggregate_declaration{
+      compute::Declaration::Sequence(
+          {std::move(input_decl),
+           {"aggregate", compute::AggregateNodeOptions{aggregates, keys, segment_keys}}}),
+      aggregate_schema};
+
+  return ProcessEmit(std::move(agg_common_opt), std::move(aggregate_declaration),
+                     std::move(aggregate_schema));
```

Comment:
`ProcessEmit` should not be part of this method.  You don't need it when working on an extension method.  This should be left back where it was in the `kAggregate` case of the top-level relation switch statement.  Then you don't need to pass in `RelCommon` and you don't need the changes to `ProcessEmit`.

Reply:
Fixed.